### PR TITLE
chore(bindings): Handle OIDC metadata changes.

### PR DESCRIPTION
**Pull request by pixlwave**
Currently when the AuthenticationService is given updated metadata, it is ignored if a dynamic registration has already been made for a selected issuer. This PR fixes that by storing the metadata's hash and resetting the store when there is a mis-match.

Additionally it moves `OidcRegistrations` out of the FFI into a new `authentication` module in the UI crate and adds some tests.

### CODE REVIEW

File: crates/matrix-sdk-ui/Cargo.toml
Comment (pixlwave) [inline crates/matrix-sdk-ui/Cargo.toml:52]: This is wrong, I need to add a feature to the crate to conditionally enable it.

File: crates/matrix-sdk-ui/src/authentication/mod.rs
Comment (pixlwave) [inline crates/matrix-sdk-ui/src/authentication/mod.rs:3]: Left this as a placeholder for refactoring the `AuthenticationService` in the FFI into lightweight bindings and moving the main logic in here.

File: crates/matrix-sdk-ui/src/authentication/mod.rs
Comment (pixlwave) [inline crates/matrix-sdk-ui/src/authentication/mod.rs:1]: So this seemed to make sense, but now running `cargo test` doesn't run the tests in here. Does CI run the tests with all features enabled?

Comment (pixlwave) [reply crates/matrix-sdk-ui/src/authentication/mod.rs:1]: > I guess the FFI crate will, in the future?

Oh I forgot to add that change in this commit 🙈

File: crates/matrix-sdk-ui/Cargo.toml
Comment (bnjbvr) [inline crates/matrix-sdk-ui/Cargo.toml:11]: As said somewhere else, we don't use Cargo features for the UI crate.

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:1]: Can you include a copyright notice above please?

File: crates/matrix-sdk-ui/src/authentication/mod.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/mod.rs:1]: Can you include a copyright notice above please?

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:74]: Since this is an internal function, it would be better to properly bubble up this error to the caller instead (i.e. have this function return a `Result<RegistrationData, ...>`).

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:89]: Could the file on disk save the entire `ClientMetadata`, or is that unsafe? Then we could use a `PartialEq` test instead of a hash comparison, which would be more robust.

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:66]: Can we hint that it does file i/o, maybe name it `read_registration_data`?

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:121]: Heh, it's a serialization error, not an invalid base path :P

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:199]: Can you also check that it doesn't affect static registrations?

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:33]: nit: please add a doc comment here too

And maybe we could make it obvious in the name that it's saved to disk, something like `FrozenRegistrationData`?

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:38]: For this field and the static one in `OidcRegistrations`, could we use wrapper types to indicate what the string represent? Maybe we could use an Url type for the key (ISTR we're using some in the code base), and for the value something like:

```rust
struct ClientId(String);
```

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:110]: Maybe `set_and_write_client_id` to hint at I/O too?

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Thanks, this is on the right path. I'm wondering if we can avoid relying on a hash and store the entire client metadata on disk instead.

File: crates/matrix-sdk-ui/Cargo.toml
Comment (jplatte) [reply crates/matrix-sdk-ui/Cargo.toml:11]: Well, not when it's just for things being experimental, at least (since the crate itself is rather unstable).

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (pixlwave) [reply crates/matrix-sdk-ui/src/authentication/oidc.rs:38]: Ah yeah, I think in strings instead of URLs in the FFI because of uniffi. That definitely makes sense.

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (pixlwave) [reply crates/matrix-sdk-ui/src/authentication/oidc.rs:89]: Yes that works, we have to validate the metadata so I've added `UnvalidatedRegistrationData` for deserialisation with a `validate()` method that returns `FrozenRegistrationData`.

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (pixlwave) [reply crates/matrix-sdk-ui/src/authentication/oidc.rs:121]: Oh, so it couldn't be a an error writing to the file?

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:51]: I think this is used indirectly for a `to_string()` call happening in the FFI; instead, since the wrapped `String` is public, could you remove this implementation, so there's no chance to do an accidental allocation, and instead callers can directly use `.0`?

File: bindings/matrix-sdk-ffi/src/authentication_service.rs
Comment (bnjbvr) [inline bindings/matrix-sdk-ffi/src/authentication_service.rs:489]: Note: this `to_string()` can be replaced by taking ownership of the underlying string:

```suggestion
            credentials: ClientCredentials::None { client_id: client_id.0 },
```

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/authentication/oidc.rs:163]: It could be that the serialization failed, OR that the writer failed; in fact it might be nice to not ignore the error in `map_err` and report it to the caller (feel free to stringify it before doing that).

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:127]: Rewinding a bit. Why is this using a file on disk, and not one of the stores? :thinking:

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:39]: nit: not hash anymore!

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:79]: Out of curiosity, what's the risk assessment behind deserializing as unvalidated? to prevent against external apps writing invalid information? or in case the validation rules change, so as to invalidate previously valid registrations?

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/authentication/oidc.rs:181]: The trace is lying, nothing's being cleared here?

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Thanks! We're getting closer :grinning:

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (pixlwave) [reply crates/matrix-sdk-ui/src/authentication/oidc.rs:127]: Because we don't have a store yet. The `AuthenticationService` does everything in memory, only setting up a Client that uses the disk right at the very end of the login request and when the service will be released.

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (pixlwave) [reply crates/matrix-sdk-ui/src/authentication/oidc.rs:79]: I'm not exactly sure on this one. [ClientMetadata](https://github.com/matrix-org/matrix-authentication-service/blob/0bee1284b247272d3eb96c73eb9b1d92151f39a8/crates/oauth2-types/src/registration/mod.rs#L129-L131) is `Deserialize` only and [VerifiedClientMetadata](https://github.com/matrix-org/matrix-authentication-service/blob/0bee1284b247272d3eb96c73eb9b1d92151f39a8/crates/oauth2-types/src/registration/mod.rs#L806-L808) is `Serialize` only. @zecakeh or @sandhose can probably provide more context on the reasons for this.

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/authentication/oidc.rs:119]: Ah right, that makes sense. Thanks.

File: crates/matrix-sdk-ui/src/authentication/oidc.rs
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/authentication/oidc.rs:71]: Ok, if that happens in a lib upstream, I suppose it's for the reasons I imagined. Curious to hear the reasoning, but let's not block merging for that.

**Review by bnjbvr — APPROVED** [review]
lgtm, thanks! Some test failures to fix before we merge, and we're all set.

**Pull request merged**
